pywalfox-native: init at 2.7.4

tsandrini opened the pull request:
## Description of changes

[pywalfox-native](https://github.com/Frewacom/pywalfox-native) is the native messaging host for the [Pywalfox](https://github.com/Frewacom/pywalfox) dynamic themer that dynamically sets up/changes themes of Firefox & Thunderbird based on themes generated from [pywal](https://github.com/dylanaraps/pywal) (which is in nixpkgs).

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

tsandrini committed:
pywalfox-native: init at 2.7.4

pywalfox: refactor sha256->hash

pywalfox: convert hash to SRI format

pywalfox-native: init at 2.7.4

pywalfox-native: clean redundancy, del maintainer

kirillrdy commented on pkgs/development/python-modules/pywalfox/default.nix:
```diff
@@ -0,0 +1,25 @@
+
+    src = fetchPypi {
+      inherit pname version;
+      sha256 = "59e73d7e27389574fb801634e03d8471f09bfe062865cad803f68c456680ed66";
```
```suggestion
      hash = "59e73d7e27389574fb801634e03d8471f09bfe062865cad803f68c456680ed66";
```

tsandrini replied:
Updated with the proper SRI format, thanks for the feedback again! :+1:

kirillrdy commented:
@tsandrini  greetings !
could you please squash at least last two commits, or even all three. This is done so that all commits  evaluate, (helps with automatic bisecting )

tsandrini commented:
> @tsandrini greetings ! could you please squash at least last two commits, or even all three. This is done so that all commits evaluate, (helps with automatic bisecting )

Hi! Thanks for the review! I completely missed the SRI format so I was looked how it works and such, thanks for the patience! I've squashed and rebased the commits into one as expected :)

kirillrdy commented on pkgs/development/python-modules/pywalfox/default.nix:
```diff
@@ -0,0 +1,25 @@
+{ lib, buildPythonPackage, fetchPypi }:
+
+buildPythonPackage rec {
```
my understanding is pywalfox is an application an not intended to be imported as python library

if that's the case, move this out of python-modules into pkgs/by-name

here is some more information https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md

also instead of using `buildPythonPackage`, use `buildPythonApplication`

tsandrini replied:
Yes, you are correct. To be fair I wasn't and I still am not sure about moving it out of the python module namespace due to the naming of the packages by the author. What I mean by that is that

- the python name for the native messenger is `pywalfox`
- which is also the official name for the firefox addon

And since the nixpkgs naming convention is to keep the pypi/python package name I thought that atleast keeping it in the python namespace makes it clear that **this is the native messaging host**.

However, one counterargument might be that the probability of Pywalfox (the ff/thun/ addon) being in nixpkgs is low since most of the people using it get it via home-manager and nur....

If you think it's the correct choice I'll move it ofc :+1:

kirillrdy replied:
- only python packages that intended to be imported in python are placed in python-modules.
- nixpkgs is currently in process  of migrating from all-packages.nix into by-name structure, therefore all new packages ( except some things  like python modules ) go there.

as for name clash that you are concerned about, In nixpkgs we can choose any name for it that makes sense.

i this case more correct name for package would be `pywalfox-native` ? ( opened for suggestions )

kirillrdy replied:
> A possible solution might be to name the package `pywalfox-native` maybe? But we can leave it at `pywalfox` as well I guess.

worth noting that nixpkgs packags kinda have two different places where packages are given name.

one os `pname`, package name attribute in derivation, this is what tools like nix-env show when you install pacakges

second name is name of attribute in nixpkgs that reefers to the package, eg nix-env -iA nixpkgs.packagename

I propose to change both to `pywalfox-native`

tsandrini replied:
Thanks! I agree! I cleaned all of the redundancy, renamed the commit & PR and separated the maintainers addition into a separate commit. Thanks a lot for the patience. Hopefully I'll be able to return it to the community, thanks :sweat_smile:

kirillrdy replied:
@tsandrini can you swap order of commits, in current order `pywalfox-native: init at 2.7.4` references maintainer that doesn't exist

tsandrini replied:
@kirillrdy I apologize for the wrong order! Should be okay now. :pray:

AndersonTorres commented on maintainers/maintainer-list.nix:
```diff
@@ -18484,6 +18484,12 @@
     github = "Trundle";
     githubId = 332418;
   };
+  tsandrini = {
```
Split this in a separate commit (no need for a new PR).

AndersonTorres commented on pkgs/development/python-modules/pywalfox/default.nix:
```diff
@@ -0,0 +1,25 @@
+
+    meta = with lib; {
+      homepage = "https://github.com/Frewacom/pywalfox-native";
+      description = "Native app used alongside the Pywalfox addon.";
```
```suggestion
      description = "Native app used alongside the Pywalfox addon";
```

AndersonTorres commented on pkgs/development/python-modules/pywalfox/default.nix:
```diff
@@ -0,0 +1,25 @@
+    meta = with lib; {
+      homepage = "https://github.com/Frewacom/pywalfox-native";
+      description = "Native app used alongside the Pywalfox addon.";
+      longDescription = ''
+        The native messaging host for the Firefox & Thunderbird addon Pywalfox.
+        (https://github.com/Frewacom/pywalfox)
+      '';
```
```suggestion
```

Redundant and therefore unneeded

AndersonTorres commented on pkgs/top-level/python-packages.nix:
```diff
@@ -11859,6 +11859,8 @@ self: super: with self; {
 
   pywal = callPackage ../development/python-modules/pywal { };
 
+  pywalfox = callPackage ../development/python-modules/pywalfox { };
```
As said before, use the by-name hierarchy.

tsandrini committed:
maintainers: add tsandrini

kirillrdy commented on pkgs/by-name/py/pywalfox-native/package.nix:
```diff
@@ -0,0 +1,22 @@
+      description = "Native app used alongside the Pywalfox addon";
+      mainProgram = "pywalfox";
+      license = licenses.mpl20;
+      maintainers = with maintainers; [ tsandrini ];
```
please add `mainProgram = "pywalfox";`

tsandrini replied:
> please add `mainProgram = "pywalfox";`

It is there? Or? I was testing it via flakes (`nix run .#pywalfox-native`) or am I missing something else?

kirillrdy replied:
```
[nix-shell:~/.cache/nixpkgs-review/pr-271023-3]$ ls ./results/pywalfox-native/bin/pywalfox 
./results/pywalfox-native/bin/pywalfox

```
https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md#deprecations-sec-release-2311-lib-deprecations

tsandrini replied:
@kirillrdy I'm sorry, I don't really understand what you are trying to say? `meta.mainProgram` has been set since the initial commit and I haven't changed that line since, running for example `nix run .#pywalfox-native light` works as expected. Also parsing `lib.getExe` on the derivation from REPL works.

kirillrdy replied:
ah, sorry mistake is on my part, i didn't see mainProgram :-)

tsandrini replied:
Ahh, no problem, I'm grateful for the patience :smile:

kirillrdy reviewed (approved):
Result of `nixpkgs-review pr 271023` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>pywalfox-native</li>
    <li>pywalfox-native.dist</li>
  </ul>
</details>

emilytrau reviewed (approved):
Welcome and thanks for contributing!

AndersonTorres commented on pkgs/by-name/py/pywalfox-native/package.nix:
```diff
@@ -0,0 +1,22 @@
+{ lib, python3, fetchPypi }:
+
+python3.pkgs.buildPythonApplication rec {
+    pname = "pywalfox-native";
```
The indentation is usually two spaces.
Why it wasn't noticed before?

tsandrini replied:
@AndersonTorres I apologize for the inconvenience. For my other nix projects I usually setup devshells with alejandra and other lints (deadnix, statix, nil, ...) with direnv linked straight to my emacs, however, this doesn't fit CONTRIBUTING.md so I was checking it manually which is obviously error prone. To be honest I kind of hoped that there would be either pre-commit checks or CI checks for the code styleguide which would pick up any of my possible style mismatches.

Anyway, I apologize once again and should I create a new PR or how should I handle this situation? :pray:

AndersonTorres replied:
We still didn't decide about code style automation. There is a recent RFC about this, tho.

The most we have is editorconfig, that checks for the typical invisible whitespaces in endlines.

For now, just open a new PR. But don't worry too much.

tsandrini replied:
Oh okay, thank you. Just a bit disappointed in myself that I tested stuff and there are still issues, but thats fine :smile: I created a simple devshell with a bunch of pre-commit checks that will hopefully catch this stuff in the future.

`pywalfox-init: refactor` is fine? :pray:

tsandrini replied:
@AndersonTorres I've created the fixup PR #271681. I was, however, confused at first that ofborg fails with 

`error: cannot coerce a Boolean to a string`

but it seems that this happens to every PR that isnt an init or update? Atleast from what I see in other PRs. The same output happens for `nixpkgs-review pr 271681`. Hopefully this is normal behaviour I guess.

AndersonTorres replied:
This error is happening all around today. Maybe some experimental code broke elsewhere.

Let's wait one day.
